// dmx-ts/dist/src/drivers/abstract-serial-driver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractSerialDriver = void 0;
const events_1 = require("events");
class AbstractSerialDriver extends events_1.EventEmitter {
    constructor(serialPort, args) {
        super();
        this._intervalHandle = undefined;
        this._sendInterval = args.sendInterval;
        this._serialPortName = serialPort;
        this._serialPortOptions = args.serialPortOptions;
        this._serialPort = serialPort;
        this._universe = Buffer.alloc(513);
        console.log('abstract serial driver; CONSTRUCTOR; serialport then args:');
        console.log(serialPort);
        console.log(args);
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('abstract serial driver; METHOD: async init(); commence; try .start()');
            try {
                console.log('.start() ing');
                // Start the driver (e.g., start sending/receiving data)
                this.start();
            }
            catch (err) {
                console.log('abstract serial driver; ERROR: could not .start()');
                console.error('Error details:', err);
                throw err;
            }
        });
    }
    ;
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            this.stop();
            yield this._serialPort.close();
        });
    }
    get serialPort() {
        return this._serialPort;
    }
    get universeBuffer() {
        return this._universe;
    }
    start() {
        if (this._intervalHandle !== undefined) {
            throw new Error('Driver is already running.');
        }
        this._intervalHandle = setInterval(this.sendUniverse.bind(this), this._sendInterval);
    }
    stop() {
        if (this._intervalHandle !== undefined) {
            clearInterval(this._intervalHandle);
            this._intervalHandle = undefined;
        }
    }
    get(channel) {
        return this._universe[channel];
    }
    update(channels, extraData) {
        for (const c in channels) {
            this._universe[c] = channels[c];
        }
        this.emit('update', channels, extraData);
    }
    updateAll(value) {
        for (let i = 1; i <= 512; i++) {
            this._universe[i] = value;
        }
    }
}
exports.AbstractSerialDriver = AbstractSerialDriver;
